Drivers: add properties files support

bjarki-andreasen opened the pull request:
**drivers: Added properties file parser and macro generator**

**Purpose**
Currently, the metadata available to the user and application for devices mainly
come from the devicetree. The devicetree primarily describes a devices hardware
properties, like baudrate, i2c addresses etc. These properties are contained in YAML
files in the dts/bindings/ folders.

This PR adds YAML files to drivers/properties/ which describe software, metadata
and zephyr specific device properties, like which API a device driver supports.

**API property**
```
compatible: "st,lis2dh"
api: "sensor"
```

The bindings files (hardware properties) YAML files are used to generate code which
is used by the application and drivers, likewise, the properties files can be used to
generate code.

An example macro generated using output from the bindings and properties files is
**DEVICE_DT_API_FOREACH(fn, api_type)**

This macro invokes fn for each node which supports the specified API type, passing
the node_id as the single argument. This allows for shells and other subsystems to
identify relevant devices at compile time, negating the need for device_get_binding()
in some cases.

```
#define SENSOR_INFO_INITIALIZER(node_id)                                \
	{                                                                   \
		.dev = DEVICE_DT_GET(node_id),                                  \
		.vendor = DEVICE_DT_PROPERTY(node_id, vendor),                  \
		.model = DEVICE_DT_PROPERTY(node_id, model),                    \
	},

static struct sensor_info sensors[] = {
	DEVICE_DT_API_FOREACH(SENSOR_INFO_INITIALIZER, sensor)
};

```

**Template properties files**
The properties files also contain the top level key "properties". This feature is used
to create template properties files for types of devices. A sensor for example must
have a vendor and model specified, and must implement the sensor API.

The template sensor-device.yaml file specifies the required and common properties:
```
properties:
  vendor:
    type: string
    required: true
  model:
    type: string
    required: true

api: sensor
```

The specific sensor includes the template and specifies the required properties:
```
include: sensor-device.yaml
compatible: "st,lis2dh"
properties:
  vendor:
    value: "STMicroelectronics"

  model:
    value: "LIS2DH"
```


**Validate and parse properties files**
This PR brings the script gen_device_properties.py
This script finds properties files, sanity checks them,
sorts them by include (properties files can include
other properties files), and then processes all supported
keys, generating useful output which can be imported by
downstream scripts.

**Scalable top level keys support**
The script contains a "user" section which is used to add
support for new top level keys. A guide for how to add
validation and parsing for a new key is contained within
this section.

**Generating macros from properties and devicetree**
This PR also brings one of said downstream scripts
gen_device_header.py This script uses the clean output from
gen_device_properties.py and the output from edtlib.py, which
contains devicetree data. The script combines data from
devicetree and properties files to generate macros, which are
used by new macros in device.h

**Update to device.h to include properties macros**
device.h has a new section added with macros for the initial
supported properties, like DEVICE_DT_PROPERTY() and
DEVICE_DT_API_FOREACH().

henrikbrixandersen commented:
The PR description just describes what has been added, not why. Please add an example use-case for this new infrastructure.

bjarki-trackunit commented:
> The PR description just describes what has been added, not why. Please add an example use-case for this new infrastructure.

Updated PR description

Commit:
drivers: Added properties file parser and macro generator

The purpose of this PR is to add support for device driver
metadata which may be used to generate code. The metadata
is stored in YAML files.

The contents of the YAML files are validated and parsed by
a script. The parsed and validated output can be imported by
downstream scripts to generate macros etc.

This PR brings the script gen_device_properties.py
This script finds properties files, sanity checks them,
sorts them by include (properties files can include
other properties files), and then processes all supported
keys, generating useful output which can be imported by
downstream scripts.

The script contains a "user" section which is used to add
support for new top level keys. A guide for how to add
validation and parsing for a new key is contained within
this section.

This PR also brings one of said downstream scripts
gen_device_header.py This script uses the clean output from
gen_device_properties.py and the output from edtlib.py, which
contains devicetree data. The script combines data from
devicetree and properties files to generate macros, which are
used by new macros in device.h

device.h has a new section added with macros for the initial
supported properties, like DEVICE_DT_PROPERTY() and
DEVICE_DT_API_FOREACH().

Signed-off-by: Bjarki Arge Andreasen <bjarkix123@gmail.com>

Commit:
drivers: Add sensor shell demo using properties files

This commit adds an demonstration of using the properties
files and macros to listify all sensors at compile time,
and get some device node specific const properties like model
and vendor.

Signed-off-by: Bjarki Arge Andreasen <bjarkix123@gmail.com>

gmarull commented on drivers/properties/sensor/sensor-device.yaml:
```diff
@@ -0,0 +1,10 @@
+properties:
+  vendor:
+    type: string
+    required: true
+
+  model:
+    type: string
+    required: true
```
both properties can be obtained from DT. From dt spec `compatible`:

> The recommended format is "manufacturer,model", where manufacturer is a string describing
the name of the manufacturer (such as a stock ticker symbol), and model specifies the model number.

Human readable string for manufacturers is found in `dts/bindings/vendor-prefixes.txt`. Summarizing, these properties already exist. There's actually ongoing work to expose them https://github.com/zephyrproject-rtos/zephyr/pull/50015

bjarki-trackunit replied:
I used these properties for the example based on this PR https://github.com/zephyrproject-rtos/zephyr/pull/49294 which intends to add these values.

The edtlib can be expanded further to also use the `dts/bindings/vendor-prefixes.txt` and compatible to look up the full name and generate a macro for it, but is that really a job for the edtlib?

The properties files should be used for immutable base properties for devices, the number of channels a PWM controller contains, the number of alarms an RTC supports, the specific data a sensor can return etc. These are base properties that are immutable, like vendor and model, which perfectly fit into the purpose of the properties files. With vendor and model however, it just happens it can be extracted from the compatible property.

I think both methods of getting vendor and model are fine, I prefer moving them to the properties files because they are designed for it, it requires more processing to get it from the edtlib.

henrikbrixandersen replied:
> The properties files should be used for immutable base properties for devices, the number of channels a PWM controller contains, the number of alarms an RTC supports, the specific data a sensor can return etc. These are base properties that are immutable, like vendor and model, which perfectly fit into the purpose of the properties files. 

Beware that the properties you describe above may not be "immutable" or rather constant across the devices supported by a given driver. Examples off the top of my head:
- The NXP MCUX FlexCAN controller driver supports the FlexCAN IP core on multiple different SoCs, but the number of available mailboxes vary from SoC to SoC.
- The TI LMP90xxx ADC driver supports an entire family of ADCs, but the resolution, number of channels etc. vary across the family.
- Same goes the the Microchip MCP320x ADC driver.

I am sure the above is true for many other in-tree drivers, e.g. many sensor drivers support multiple variants of a given sensor, each with different properties.

henrikbrixandersen replied:
> The properties files should be used for immutable base properties for devices, the number of channels a PWM controller contains, the number of alarms an RTC supports, the specific data a sensor can return etc.

These example properties, I might add, sound more like a hardware description than a driver description and properly belong in the devicetree.

bjarki-trackunit replied:
Sure, properties files are not a good solution for hardware model revision or layout specific properties. The are useful for information regarding a device driver and its capabilities, like which specific API it supports, which features it supports, which hardware revisions it supports etc.

galak reviewed (commented):
Can you look to add a pykwalify yaml that describes the format of the property yaml files.

galak commented on drivers/properties/serial/zephyr,native-posix-uart.yaml:
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2022 Trackunit A/S
```
Seems like we should have `uart.yaml` specify `api: uart` and include `uart.yaml` in driver specific YAMLs

galak replied:
Would also like to see how we specify things like standard uart features, like does this driver support async or not.

bjarki-trackunit replied:
The perfect way of doing this would be to have multiple APIs for it
`api: ["uart_poll", "uart_async", "uart_interrupt"]`

bit since that is not possible right now, the uart.yaml would contain:
```
properties:
  supports_uart_poll:
    type: bool
  supports_uart_async:
    type: bool
  supports_uart_interrupt:
    type: bool 
```

The specific uart file would then look like this:
```
include: "uart.yaml"

compatible: "dummy,uart"

properties:
  supports_uart_poll:
    value: true
```

This is just an example, we can easily add new keys and features to the files if we find a better way of doing it

bjarki-trackunit replied:
I have a better idea:

Lets add a key named features:

uart.yaml:
```
features:
    define:
      - uart_poll
      - uart_interrupt
      - uart_async
```

and dummy,uart.yaml
```
include: "uart.yaml"

compatible: "dummy,uart"

features:
  implements:
    - uart_poll
    - uart_interrupt
```

which could be tested for using
```
DEVICE_DT_FEATURE_SUPPORTED(node_id, uart_poll)
```

bjarki-trackunit replied:
Better idea yet

uart.yaml:
```
api:
  uart:
    features:
      - poll
      - interrupt
      - async
```

and dummy,uart.yaml
```
include: "uart.yaml"

compatible: "dummy,uart"

api:
  uart:
    implements:
      - poll
      - interrupt
```

and to test for features
```
DEVICE_DT_API_IMPLEMENTS(node_id, uart, poll)
```

gmarull replied:
> The perfect way of doing this would be to have multiple APIs for it `api: ["uart_poll", "uart_async", "uart_interrupt"]`
> 
> bit since that is not possible right now, the uart.yaml would contain:
> 
> ```
> properties:
>   supports_uart_poll:
>     type: bool
>     required: true
>   supports_uart_async:
>     type: bool
>     required: true
>   supports_uart_interrupt:
>     type: bool
>     required: true
> ```
> 
> The specific uart file would then look like this:
> 
> ```
> include: "uart.yaml"
> 
> compatible: "dummy,uart"
> 
> properties:
>   supports_uart_poll:
>     value: true
>   supports_uart_interrupt:
>     value: true
>   supports_uart_async:
>     value: false
> ```
> 
> Which can be determined at compile time using
> 
> ```
> #if DEVICE_DT_PROPERTY(node_id, supports_uart_interrupt)
> #endif
> ```
> 
> **Note!** that they properties dont support booleans yet, i will add this shortly, I forgot about that one :)

We have Kconfig to specify wether a serial driver supports async/interrupt: `SERIAL_SUPPORT_ASYNC`, `SERIAL_SUPPORT_INTERRUPT`. Why do we need them in another file now?

galak commented on drivers/properties/sensor/sensor-device.yaml:
```diff
@@ -0,0 +1,10 @@
+properties:
+  vendor:
```
I don't think vendor & model make sense here since we can extract them from devicetree.

bjarki-trackunit replied:
I added my opinion to this as part of previous comment :)

galak commented:
@tejlmand would love your input on this and how we might integrate this week the build system (Kconfig, Cmake) from a consistency point of view, as well if we could or should move some of this data into one of their existing systems.

galak commented:
Add links to some discussion PRs around the topic of file metadata:

https://github.com/zephyrproject-rtos/zephyr/issues/49431
https://github.com/zephyrproject-rtos/zephyr/issues/38725#issuecomment-1055570100

galak commented:
a few questions:
* How do we associate this information with the file(s), kconfig symbols, and devicetree compatible?
* How do we ensure this information is set and consistent with the actual code.  What kinda build asserts might we add? checks in cmake?

bjarki-trackunit commented:
To clarify what separates the bindings files from the properties files:

**Device bindings**
- Are templates for hardware
- Declare properties which are set in devicetree

**Properties files**
- Are templates for software
- Declare and set properties

This is just a note for tomorrow :)

bjarki-trackunit commented:
> a few questions:
> 
>     * How do we associate this information with the file(s), kconfig symbols, and devicetree compatible?

These files use the same compatible as the devicetree, so both bindings and properties files contain the key compatible: "st,lis2dh" for example. The properties are not tied to the Kconfigs at all currently.

>     * How do we ensure this information is set and consistent with the actual code.  What kinda build asserts might we add? checks in cmake?

The properties files are used to generates a dictionary with the properties which can be used by scripts to for example creating macros (like the gen_device_header.py does) and possibly validating coherency between code and properties. What and how these scripts would validate coherency i don't know :)

bjarki-trackunit commented:
> Can you look to add a pykwalify yaml that describes the format of the property yaml files.

I will look that up, and most likely use it, could possibly remove a lot of "assert isInstance(...)" etc. from the gen_device_properties.py

henrikbrixandersen commented:
@bjarki-trackunit I am sorry, but I still do not fully understand the use-case for these property files. The PR explains the "how" part but I have yet to understand the "why" part. If you can, please present a couple of actual use-cases for this in the dev review later today - or document them here.

henrikbrixandersen commented:
> @henrikbrixandersen The use case for the properties files is to describe the capabilities of hardware from a user/application perspective, which is through the capabilties of the device drivers, which are described by the properties files.

Again, you are describing the "how" part, not the "why" part. The "why" part should cover the nature of the problem are you trying to solve.

On another note, your example of various sensor types listed in the base sensor YAML file does not address out-of-tree sensor types.

bjarki-trackunit commented:
> > @henrikbrixandersen The use case for the properties files is to describe the capabilities of hardware from a user/application perspective, which is through the capabilties of the device drivers, which are described by the properties files.
> 
> Again, you are describing the "how" part, not the "why" part. The "why" part should cover the nature of the problem are you trying to solve.
> 
> On another note, your example of various sensor types listed in the base sensor YAML file does not address out-of-tree sensor types.

It is not obvious without looking at the code, but just like dts files which can be added out-of-tree in /dts/bindings, properties files can be added in /drivers/properties.

We need metadata regarding device drivers which can be used by build system and CI. We need this to describe the APIs supported by a driver, which is required for the multi API device model and shell subsystem. Thats the why.

henrikbrixandersen commented:
The proposed property files are to store meta information about a given driver, linked by the DTS compatible string. The discussions so far, at least the discussions I have taken part in, have been concentrated around the “how” part of storing driver meta information but seems to have skipped more lightly around formalising the “why” part. Below I have tried to capture some of the “why” arguments I have picked up during the last few discussions on this topic. Beware that the wording is my own (and I may have misunderstood the reasoning behind several of these bullets, so please correct me if I have).

I have included my own suggestion of alternative proposals of how to solve these problems without the use of property files as I have yet to be convinced we have a need for them.

- **Problem:** Who maintains this driver? We could store maintainer information in the property files. 
  - **Alternative proposal:** Since the proposed property files are to be linked to the driver using the same compatible string as in the DTS binding we might as well store the maintainer information in the DTS binding YAML file. However, storing the driver maintainer information in any of these files doesn’t buy us much, as there is no linking from the compatible string to the file(s) which the driver consists of. Furthermore, we tend to group maintainers (for instance a group of maintainers are responsible for all STM32 drivers). Spreading this information out into several binding or property files will make this much harder to maintain (each STM32 property file would have to be updated to list a new STM32 maintainer). We need a pattern based approach as in current MAINTAINERS file. Furthermore, we have many files in the tree that does not belong to a driver/compatible. We would need at least two different ways of keeping track of maintainers (e.g. MAINTAINERS file for non-driver files + property files for driver files).

- **Problem:** My application/library depends on having a specific type of sensor with a given sample frequency/update rate. We could store the sensor type and resolutions/sample frequency/… in the property file and have compilation fail if such a sensor is not present in the board being built for. 
  - **Alternative proposal:** Properties like sample frequency, resolution, sensor type etc. are all properties of the hardware, not the driver. They should go in the devicetree (perhaps as const or default values in the binding if they are immutable). An application depending on a specific type of sensor/other device can use a DTS alias for obtaining a reference to that particular device. This would either require the use of generic aliases or application-specific aliases. It is easy to test for the present of such an alias being present and enabled during compile time using build assertions and the DT_* macros. If the needed hardware properties are present in the DTS binding (as described previously), it is also easy to test for these and fail compilation if they don’t meet the application requirements. 

- **Problem:** My application needs to know whether a given UART implements the polling, interrupt, or asynchronous parts of the UART API. We could store which part of the UART API is implemented by a given UART driver in its property file. 
  - **Alternative proposal:** The UART API is a patchwork of different APIs rolled into one. Instead of coming up with ways to work around this patchwork of an API, I think it may be time to clean up the UART API instead, perhaps splitting it into multiple APIs (which is effectively what the proposal here is trying to do, split the UART API in three, but keep them in the same API struct).

- **Problem:** I need to iterate through all devices of a given device class (e.g. all devices implementing the sensor API) at run time. We could generate C code macros for iterating based on the information in the property files.
  - **Alternative proposal:** The CAN and I2C driver classes already support class-specific device initialisation macros. This design pattern could be expanded to other driver classes as well and used for registering the type of driver being initialised (meaning the type of API the given struct device implements). The type information could be used for iterating a given class of devices at run time.

- **Problem:** We need to be able to generate better documentation of the features supported by a given driver. 
  - **Alternative proposal:** I am not sure I fully understand what is missing here. We have documentation generated from the various DTS bindings and documentation for the various driver Kconfigs, but granted - they are not linked together as it is today. Perhaps the recently introduced DT_HAS_* Kconfig symbol dependencies could be used for tying the driver Kconfig documentation to the corresponding devicetree binding documentation?

There are probably many other use-cases for driver meta-information, so please feel free to add further in reply to this post. Comments on my alternative proposals are of course also very welcome.

bjarki-trackunit commented:
The implementation in this PR is no longer relevant as this solution exceeds it https://github.com/zephyrproject-rtos/zephyr/issues/50788

mbolivar-nordic commented:
> * Perhaps the recently introduced DT_HAS_* Kconfig symbol dependencies could be used for tying the driver Kconfig documentation to the corresponding devicetree binding documentation?

I've tried to do this by adding doxygen comments in the drivers themselves listing the DT compatibles and other metadata associated with the driver. The doxygen approach ultimately failed because we do too much macro magic in our C files, and I kept having to update the PREDEFINED key in the doxygen configuration file even just inside my PR while rebasing.

I came to the conclusion that a separate file (or, more likely, files) listing drivers (probably by subsystem) along with associated metadata about them (for me, this metadata is: Kconfig symbols, DT compatibles, API subsets supported, maintainers, information about associated DT binding C macros to use when writing overlays, extensions to the "base" API proprietary to the driver and their headers, and free-form text in RST format to include in the docs).
